### ISSUE TITLE
Channel operator to flatten by one level

### ISSUE THREAD

**Issue by bskubi** [nextflow-io/nextflow#4847]
## New feature

A feature to flatten items within an element by one level.

## Usage scenario 

Example:

Let's say we have two channels, fq_ch and ref_ch:

fq_ch
```
[sample_id: "id1", reads: ["/data/s1_1.fq", "/data/s1_2.fq"]]
[sample_id: "id2", reads: ["/data/s2_1.fq", "/data/s2_2.fq"]]
```

ref_ch
```
[ref_genome: "/ref/hg38.fa"]
```

We'd like to combine them to use them as a process input. Since ref_ch only has one value, we need to use:
```
fq_ch.combine(ref_ch).set{align_ch}
```

However, this produces elements with two items, which will be hard to pass to a process as input since processes don't accept tuples of tuples.
```
[[sample_id: "id1", reads: ["/data/s1_1.fq", "/data/s1_2.fq"]], [ref_genome: "/ref/hg38.fa"]]
[[sample_id: "id2", reads: ["/data/s2_1.fq", "/data/s2_2.fq"]], [ref_genome: "/ref/hg38.fa"]]
```

We therefore need to call:
```
fq_ch.combine(ref_ch).map{it[0] + it[1]}.set{align_ch}
```

This flattens the elements by a single step:
```
[sample_id: "id1", reads: ["/data/s1_1.fq", "/data/s1_2.fq"], ref_genome: "/ref/hg38.fa"]
[sample_id: "id2", reads: ["/data/s2_1.fq", "/data/s2_2.fq"], ref_genome: "/ref/hg38.fa"]
```

Yielding an element format we can input to a process as a single three-item tuple.

It is not immediately what the workflow is doing or why on the `map` step, particularly if `fq_ch` and `ref_ch` are created elsewhere, so that their element structure is not locally visible.

I think it would be more clear if Nextflow provided a channel operation to accomplish this, or perhaps added an optional argument to the existing `flatten` to specify the number of levels of flattening to do.

## Suggest implementation 

Here's an example implementation (I'm no Groovy expert, this is ChatGPT's port of my python code) of a flattenByOne function. You could have the channel operator call it the specified number of flattening levels on each channel element.

```
def flattenByOne(Iterable it) {
    def result = [:]
    it.each { i ->
        if (i instanceof Iterable) {
            result += i // Concatenate if it's an Iterable
        } else {
            result << i // Append if it's not an Iterable
        }
    }
    return result
}
```

**Comment by bentsherman** [comment]
I think `flatMap` will do this for you. If you don't provide a closure, it should default to `{ it -> it }`

**Comment by bskubi** [comment]
It unfortunately doesn't. If each element is a list of three hashmaps, flatMap() emits each hashmap as a separate output. The desired behavior is for it to combine the three hashmaps into a single hashmap and emit that as an output.

**Comment by robsyme** [comment]
The trick is that I think Benjamin (bskubi) isn't strictly looking for a flatten operation, I think (correct me if I'm wrong) they are looking for a `map` + `plus` operation.

If readability is a concern, perhaps defining the arguments to the closure will make the operation a little more clear. You could use a `map` operation with named arguments 

```groovy
map { fqMap, refMap -> fqMap + refMap }
```

In context:
```groovy
workflow {
    fq_ch = Channel.of(
        [sample_id: "id1", reads: ["/data/s1_1.fq", "/data/s1_2.fq"]],
        [sample_id: "id2", reads: ["/data/s2_1.fq", "/data/s2_2.fq"]]
    )

    ref_ch = Channel.of([ref_genome: "/ref/hg38.fa"])

    fq_ch
    | combine(ref_ch)
    | map { fqMap, refMap -> fqMap + refMap }
    | flatMap
    | view
}
```

Would this satisfy the readabilitiy concerns, @bskubi?

**Comment by robsyme** [comment]
If you want to make the method really flexible for an unknown number of Maps, you could use the `inject` method defined on collections ([docs](https://docs.groovy-lang.org/latest/html/groovy-jdk/java/util/Collection.html#inject(java.lang.Object,%20groovy.lang.Closure)), [helpful tutorial](https://blog.mrhaki.com/2009/09/groovy-goodness-using-inject-method.html)):

```groovy
workflow {
    fq_ch = Channel.of(
        [sample_id: "id1", reads: ["/data/s1_1.fq", "/data/s1_2.fq"]],
        [sample_id: "id2", reads: ["/data/s2_1.fq", "/data/s2_2.fq"]]
    )

    ref_ch = Channel.of([ref_genome: "/ref/hg38.fa"])

    fq_ch
    | combine(ref_ch)
    | map { it.inject([:]) { result, map -> result + map } }
    | view
}
```

If you find yourself using this closure multiple times in your workflow, you could use a named closure:

```groovy
reduceCollectionOfMaps = { it.inject([:]) { result, map -> result + map } }

workflow {
    fq_ch = Channel.of(
        [sample_id: "id1", reads: ["/data/s1_1.fq", "/data/s1_2.fq"]],
        [sample_id: "id2", reads: ["/data/s2_1.fq", "/data/s2_2.fq"]]
    )

    ref_ch = Channel.of([ref_genome: "/ref/hg38.fa"])

    fq_ch
    | combine(ref_ch)
    | map(reduceCollectionOfMaps)
    | view
}
```

**Comment by bskubi** [comment]
@robsyme You're correctly understanding my goal! I think the version with the named closure is best from a readability and maintainability standpoint given that Nextflow doesn't provide a standard operator for this purpose. Thank you!

**Comment by bskubi** [comment]
Actually, it seems named closures can't be included across modules so I would probably have to adapt it as a function.

**Comment by robsyme** [comment]
Great! I'd argue that it using the `inject` method in a map closure:

```groovy
map { it.inject(Map::plus) }
```

 is small enough that it doesn't warrant a separate function, but Nextflow _does_ support additional functions, so you should take whichever approach feels most appropriate for your codebase.

Your initial request was for a new operator to support this function. Given that the closure above is so small, a separate operator is unlikely. Given this, would you be ok if we closed this issue?

**Comment by bskubi** [comment]
@robsyme Yes, I'm fine with closing the issue. Thank you for your thoughtful replies, that was very helpful. On reflection, the best option for readability short of a separate operator is probably putting the named enclosure immediately before the channel definition that uses it so I will probably just do that.
